Per route limits (#34)
fastify/fastify-rate-limit: README.md

  keyGenerator: function(req) { /* ... */ }, // default (req) => req.raw.ip
})
```

> - `global` : will tell to the plugin if all route will be set will a `rate limit`

Reviewer: I think you mean:

```suggestion
- `global` : indicates if the plugin should apply a rate limit to all routes within the encapsulation scope
```

})
```

### Options on the endpoint itself

> You can overide the default value set during the registration of the plugin from the route declaration, the values will be used `only` for this route.

Reviewer: ```suggestion
Rate limiting can be configured on a per route basis by supplying rate limit configuration on the route declaration.
```

Regarding the `whitelist`. There are 2 types :

>  - global : will affect all the endpoint.

Reviewer: ```suggestion
 - global : will affect all endpoints within the encapsulation scope
```

>  - endpoint : will affect only the targeted endpoint.

Reviewer: ```suggestion
 - endpoint : will affect only the targeted endpoint
```

> `The Global whitelist` is set in the plugin when registering it with fastify.register(...).

Reviewer: ```suggestion
The global whitelist is configured when registering it with `fastify.register(...)`.
```

> `The Endpoint whitelist` is set on the endpoint directly with the { config : rateLimit : whitelist : [] }object.

Reviewer: ```suggestion
The endpoint whitelist is set on the endpoint directly with the `{ config : { rateLimit : { whitelist : [] } } }` object.
```

> you need to set the option `whiteListInRedis` to `true` when registering the plugin. This is only available when using redis. 

Reviewer: This seems to come out of nowhere. Don't know what to suggest here.

> The ACL is checking based on the value of the key from the keyGenerator. In this example we are checking the Ip but it could be a whitelist of specific user id.

Reviewer: ```suggestion
ACL checking is performed based on the value of the key from the `keyGenerator`. In this example we are checking the IP address, but it could be a whitelist of specific user identifiers.
```

    whitelist: ['192.168.0.10'], // global whitelist access. 
    redis: redis, // connection to redis
    skipOnError: false // default false
    // keyGenerator: function(req) { /* ... */ }, // default (req) => req.raw.ip
  })

fastify.get('/', {
  config: {
    rateLimit: {
      max: 3,
      timeWindow: '1 minute'
    }
  }
}, (req, reply) => {
  reply.send({ hello: 'from ... root' })
})

fastify.get('/private', {
  config: {
    rateLimit: {
      max: 3,
      timeWindow: '1 minute'
    }
  }
}, (req, reply) => {
  reply.send({ hello: 'from ... private' })
})

fastify.get('/public', (req, reply) => {
  reply.send({ hello: 'from ... public' })
})

fastify.get('/public/sub-rated-1', {
  config: {
    rateLimit: {
      timeWindow: '1 minute',
      onExceeding: function (req) {
        console.log('callback on exceededing ... executed before response to client')
      },
      onExceeded: function (req) {
        console.log('callback on exceeded ... to black ip in security group for example, req is give as argument')
      }
    }
  }
}, (req, reply) => {
  reply.send({ hello: 'from sub-rated-1 ... using default max value ... ' })
})
```

> - `max`: is the maximum number of requests a single client can perform inside a timeWindow.

Reviewer: ```suggestion
- `max`: the maximum number of requests a single client can perform inside a `timeWindow`.
```

> - `timeWindow:` the duration of the time window, can be expressed in milliseconds (as a number) or as a string, see [`ms`](https://github.com/zeit/ms) too see the supported formats.

Reviewer: ```suggestion
- `timeWindow:` the duration of the time window, can be expressed in milliseconds (as a number) or as a string (see [`ms`](https://github.com/zeit/ms) for supported formats).
```

> - `onExceeding` : `callback` that will be executed each time that a user goes via a route which is `rate limited`

Reviewer: ```suggestion
- `onExceeding` : callback that will be executed each time a request is made to a route that is rate limited.
```

> - `onExceeded` : `callback` that will be executed when a user reached the maximum number of tries. Can be useful to blacklist the ip directly in a securityGroup if on aws using the api.

Reviewer: ```suggestion
- `onExceeded` : callback that will be executed when a user reached the maximum number of tries. Can be useful to blacklist clients.
```

> - `keyGenerator`: a function to generate a unique identifier for each incoming request. Defaults to `(req) => req.ip`, the IP is resolved by fastify using `req.connection.remoteAddress` or `req.headers['x-forwarded-for']` if [trustProxy](https://www.fastify.io/docs/master/Server/#trustproxy) option is enabled. Use it if you want to override this behavior.

Reviewer: ```suggestion
- `keyGenerator`: a function to generate a unique identifier for each incoming request. Defaults to `(req) => req.ip`, the IP is resolved by Fastify using `req.connection.remoteAddress` or `req.headers['x-forwarded-for']` if the [`trustProxy`](https://www.fastify.io/docs/master/Server/#trustproxy) option is enabled. Specify this option to override this behavior.
```
